File: ui/src/api/hooks/cashflow.ts
import { useQuery } from "@tanstack/react-query";
import { commands, type CashflowForecast } from "../client";
import { isBackendAvailable } from "../../utils/runtime";

export interface CashflowParams {
  horizonDays: number;
  bufferCents: number;
  /** Hypothetical one-off spend to test; 0 = none. */
  extraExpenseCents: number;
}

/**
 * Near-term daily cash-flow forecast + safe-to-spend from the shared
 * `finsight-core::cashflow` layer. The buffer and hypothetical spend are pure
 * what-if parameters — evaluated server-side without persisting anything — so
 * changing them just refetches a fresh projection.
 */
export function useCashflowForecast(params: CashflowParams) {
  return useQuery<CashflowForecast>({
    queryKey: ["cashflow-forecast", params.horizonDays, params.bufferCents, params.extraExpenseCents],
    queryFn: async () => {
      const result = await commands.getCashflowForecast(
        params.horizonDays,
        params.bufferCents,
        params.extraExpenseCents > 0 ? params.extraExpenseCents : null,
        null,
      );
      if (result.status === "error") throw new Error(result.error.message);
      return result.data;
    },
    staleTime: 30_000,
    // Keep showing the previous forecast while a new buffer/horizon refetches,
    // so the chart doesn't flash empty on every keystroke.
    placeholderData: (prev) => prev,
    enabled: isBackendAvailable(),
  });
}

File: crates/finsight-bindings/src/commands/cashflow.rs
//! Codegen wrapper for the near-term cash-flow forecast + safe-to-spend.

use crate::error::AppResult;
use crate::AppState;
use finsight_core::cashflow::CashflowForecast;

/// Project the liquid balance forward `horizonDays` (default 30, clamped 7–90),
/// optionally against a safety buffer and a hypothetical one-off outflow — all
/// evaluated purely, nothing persisted.
#[tauri::command]
#[specta::specta]
pub async fn get_cashflow_forecast(
    state: tauri::State<'_, AppState>,
    horizon_days: Option<i64>,
    buffer_cents: Option<i64>,
    extra_expense_cents: Option<i64>,
    extra_expense_date: Option<String>,
) -> AppResult<CashflowForecast> {
    finsight_api::commands::cashflow::get_cashflow_forecast(
        &state.api,
        horizon_days,
        buffer_cents,
        extra_expense_cents,
        extra_expense_date,
    )
    .await
}
